### ISSUE TITLE
General Moodle XML preview?

### ISSUE THREAD

**Issue by alsaibie** [ethan-ou/vscode-gift#6]
Hi, 

I'm wondering if it's possible to extend this extension to also preview Moodle XML questions? Does the parser convert the gift to xml for viewing? The GIFT extension is great, but it has some limitations, so if one would generate questions in XML directly they can exploit more options. Any pointers would be appreciated.

**Comment by ethan-ou** [comment]
The parser converts GIFT to a JSON-like format, which then generates the preview from there. In order to do the same thing with XML, we'd need a parser for Moodle's XML, which apparently is quite difficult to implement.

I had some discussions with [Chris Fuhrman](https://github.com/fuhrmanator), the author of the parser for this VSCode extension and he also tried parsing Moodle XML, but found it a dead end. Moodle's XML doesn't follow a strict specification which makes it hard to build a parser for.

It would also be difficult to write questions directly into Moodle XML and I think most users at that stage would prefer to use Moodle's UI. You'll probably need to make the tradeoff at that point whether you want to partially write your questions in GIFT then fill it in Moodle, or to just stick to Moodle's UI.

In summary:

* No parser at the moment for Moodle XML (especially one that can live update such as this preview)
* Moodle XML is difficult to write in, so it wouldn't be very user friendly for you

There aren't any easy solutions for this problem, so let me know if you find anything that helps you.

**Comment by alsaibie** [comment]
Thanks for the clarification. I'm actually thinking of generating the XML via code, off a template xml, together with the gift. I already am setting up a Julia pkg to generate gift formatted questions, and plan to extend it to bare LaTeX and XML (One setup to generate homework assignments and quizzes for engineering type courses). 

So I guess the alternative is to setup an offline moodle installation and generate the xml into the local site directory and view it over the browser. I haven't tried, I hope it works and that no import process is needed.

**Comment by ethan-ou** [comment]
That sounds like an interesting idea. Especially for an engineering course, GIFT is pretty limited in what it can handle. I hope you're able to generate the questions you're looking for with your setup.

I had a look once at trying to reverse engineer Moodle's parsing code too. I think in theory it's possible to rewrite the parser for Moodle XML in Javascript and have it display in the browser, but it's a lot of work. So I think you're going in the right direction.

**Comment by alsaibie** [comment]
My planned approach is to save XML templates (eg: mcq.xml.in) for each question type, labeling them with unique placeholders, then regex replace them with the respective components of the generated question, or replace them with empty fields otherwise.

I'll try and see.
